File: server/core/src/main/java/dev/slimevr/protocol/rpc/settings/RPCSettingsHandler.java
package dev.slimevr.protocol.rpc.settings;

import com.google.flatbuffers.FlatBufferBuilder;
import dev.slimevr.bridge.ISteamVRBridge;
import dev.slimevr.config.*;
import dev.slimevr.filtering.TrackerFilters;
import dev.slimevr.osc.OSCRouter;
import dev.slimevr.osc.VMCHandler;
import dev.slimevr.osc.VRCOSCHandler;
import dev.slimevr.protocol.GenericConnection;
import dev.slimevr.protocol.ProtocolAPI;
import dev.slimevr.protocol.rpc.RPCHandler;
import dev.slimevr.tracking.processor.config.SkeletonConfigToggles;
import dev.slimevr.tracking.processor.config.SkeletonConfigValues;
import dev.slimevr.tracking.trackers.TrackerRole;
import solarxr_protocol.rpc.*;


public class RPCSettingsHandler {

	public RPCHandler rpcHandler;
	public ProtocolAPI api;

	public RPCSettingsHandler(RPCHandler rpcHandler, ProtocolAPI api) {
		this.rpcHandler = rpcHandler;
		this.api = api;

		rpcHandler.registerPacketListener(RpcMessage.SettingsRequest, this::onSettingsRequest);
		rpcHandler
			.registerPacketListener(
				RpcMessage.ChangeSettingsRequest,
				this::onChangeSettingsRequest
			);
	}

	public void onSettingsRequest(GenericConnection conn, RpcMessageHeader messageHeader) {
		FlatBufferBuilder fbb = new FlatBufferBuilder(32);

		ISteamVRBridge bridge = this.api.server.getVRBridge(ISteamVRBridge.class);

		int settings = SettingsResponse
			.createSettingsResponse(
				fbb,
				RPCSettingsBuilder.createSteamVRSettings(fbb, bridge),
				RPCSettingsBuilder
					.createFilterSettings(
						fbb,
						this.api.server.configManager.getVrConfig().getFilters()
					),
				RPCSettingsBuilder
					.createDriftCompensationSettings(
						fbb,
						this.api.server.configManager.getVrConfig().getDriftCompensation()
					),
				RPCSettingsBuilder
					.createOSCRouterSettings(
						fbb,
						this.api.server.configManager.getVrConfig().getOscRouter()
					),
				RPCSettingsBuilder
					.createVRCOSCSettings(
						fbb,
						this.api.server.configManager.getVrConfig().getVrcOSC()
					),
				RPCSettingsBuilder
					.createVMCOSCSettings(
						fbb,
						this.api.server.configManager.getVrConfig().getVMC()
					),
				RPCSettingsBuilder
					.createModelSettings(
						fbb,
						this.api.server.humanPoseManager,
						this.api.server.configManager.getVrConfig().getLegTweaks()
					),
				RPCSettingsBuilder
					.createTapDetectionSettings(
						fbb,
						this.api.server.configManager.getVrConfig().getTapDetection()
					),
				RPCSettingsBuilder
					.createAutoBoneSettings(
						fbb,
						this.api.server.configManager.getVrConfig().getAutoBone()
					),
				RPCSettingsBuilder
					.createArmsResetModeSettings(
						fbb,
						this.api.server.configManager.getVrConfig().getResetsConfig()
					)
			);
		int outbound = rpcHandler.createRPCMessage(fbb, RpcMessage.SettingsResponse, settings);
		fbb.finish(outbound);
		conn.send(fbb.dataBuffer());
	}

	public void onChangeSettingsRequest(GenericConnection conn, RpcMessageHeader messageHeader) {
		ChangeSettingsRequest req = (ChangeSettingsRequest) messageHeader
			.message(new ChangeSettingsRequest());
		if (req == null)
			return;

		if (req.steamVrTrackers() != null) {
			ISteamVRBridge bridge = this.api.server
				.getVRBridge(ISteamVRBridge.class);

			if (bridge != null) {
				bridge.changeShareSettings(TrackerRole.WAIST, req.steamVrTrackers().waist());
				bridge.changeShareSettings(TrackerRole.CHEST, req.steamVrTrackers().chest());
				bridge.changeShareSettings(TrackerRole.LEFT_FOOT, req.steamVrTrackers().feet());
				bridge.changeShareSettings(TrackerRole.RIGHT_FOOT, req.steamVrTrackers().feet());
				bridge.changeShareSettings(TrackerRole.LEFT_KNEE, req.steamVrTrackers().knees());
				bridge.changeShareSettings(TrackerRole.RIGHT_KNEE, req.steamVrTrackers().knees());
				bridge.changeShareSettings(TrackerRole.LEFT_ELBOW, req.steamVrTrackers().elbows());
				bridge.changeShareSettings(TrackerRole.RIGHT_ELBOW, req.steamVrTrackers().elbows());
				bridge.changeShareSettings(TrackerRole.LEFT_HAND, req.steamVrTrackers().hands());
				bridge.changeShareSettings(TrackerRole.RIGHT_HAND, req.steamVrTrackers().hands());
			}
		}

		if (req.filtering() != null) {
			TrackerFilters type = TrackerFilters.fromId(req.filtering().type());
			if (type != null) {
				FiltersConfig filtersConfig = this.api.server.configManager
					.getVrConfig()
					.getFilters();
				filtersConfig.setType(type.getConfigKey());
				filtersConfig.setAmount(req.filtering().amount());
				filtersConfig.updateTrackersFilters();
			}
		}

		if (req.driftCompensation() != null) {
			DriftCompensationConfig driftCompensationConfig = this.api.server.configManager
				.getVrConfig()
				.getDriftCompensation();
			driftCompensationConfig.setEnabled(req.driftCompensation().enabled());
			driftCompensationConfig.setAmount(req.driftCompensation().amount());
			driftCompensationConfig.setMaxResets(req.driftCompensation().maxResets());
			driftCompensationConfig.updateTrackersDriftCompensation();
		}

		if (req.oscRouter() != null) {
			OSCConfig oscRouterConfig = this.api.server.configManager
				.getVrConfig()
				.getOscRouter();
			if (oscRouterConfig != null) {
				OSCRouter oscRouter = this.api.server.getOSCRouter();
				var osc = req.oscRouter().oscSettings();
				if (osc != null) {
					oscRouterConfig.setEnabled(osc.enabled());
					oscRouterConfig.setPortIn(osc.portIn());
					oscRouterConfig.setPortOut(osc.portOut());
					oscRouterConfig.setAddress(osc.address());
				}

				oscRouter.refreshSettings(true);
			}
		}

		if (req.vrcOsc() != null) {
			VRCOSCConfig vrcOSCConfig = this.api.server.configManager
				.getVrConfig()
				.getVrcOSC();
			if (vrcOSCConfig != null) {
				VRCOSCHandler VRCOSCHandler = this.api.server.vrcOSCHandler;
				var osc = req.vrcOsc().oscSettings();
				var trackers = req.vrcOsc().trackers();

				if (osc != null) {
					vrcOSCConfig.setEnabled(osc.enabled());
					vrcOSCConfig.setPortIn(osc.portIn());
					vrcOSCConfig.setPortOut(osc.portOut());
					vrcOSCConfig.setAddress(osc.address());
				}
				if (trackers != null) {
					vrcOSCConfig.setOSCTrackerRole(TrackerRole.HEAD, trackers.head());
					vrcOSCConfig.setOSCTrackerRole(TrackerRole.CHEST, trackers.chest());
					vrcOSCConfig.setOSCTrackerRole(TrackerRole.WAIST, trackers.waist());
					vrcOSCConfig.setOSCTrackerRole(TrackerRole.LEFT_KNEE, trackers.knees());
					vrcOSCConfig.setOSCTrackerRole(TrackerRole.RIGHT_KNEE, trackers.knees());
					vrcOSCConfig.setOSCTrackerRole(TrackerRole.LEFT_FOOT, trackers.feet());
					vrcOSCConfig.setOSCTrackerRole(TrackerRole.RIGHT_FOOT, trackers.feet());
					vrcOSCConfig.setOSCTrackerRole(TrackerRole.LEFT_ELBOW, trackers.elbows());
					vrcOSCConfig.setOSCTrackerRole(TrackerRole.RIGHT_ELBOW, trackers.elbows());
					vrcOSCConfig.setOSCTrackerRole(TrackerRole.LEFT_HAND, trackers.hands());
					vrcOSCConfig.setOSCTrackerRole(TrackerRole.RIGHT_HAND, trackers.hands());
				}

				VRCOSCHandler.refreshSettings(true);
			}
		}

		if (req.vmcOsc() != null) {
			VMCConfig vmcConfig = this.api.server.configManager
				.getVrConfig()
				.getVMC();
			if (vmcConfig != null) {
				VMCHandler VMCHandler = this.api.server.getVMCHandler();
				var osc = req.vmcOsc().oscSettings();

				if (osc != null) {
					vmcConfig.setEnabled(osc.enabled());
					vmcConfig.setPortIn(osc.portIn());
					vmcConfig.setPortOut(osc.portOut());
					vmcConfig.setAddress(osc.address());
				}
				if (req.vmcOsc().vrmJson() != null)
					vmcConfig.setVrmJson(req.vmcOsc().vrmJson());
				vmcConfig.setAnchorHip(req.vmcOsc().anchorHip());

				VMCHandler.refreshSettings(true);
			}
		}

		if (req.tapDetectionSettings() != null) {
			TapDetectionConfig tapDetectionConfig = this.api.server.configManager
				.getVrConfig()
				.getTapDetection();
			var tapDetectionSettings = req.tapDetectionSettings();

			if (tapDetectionSettings != null) {
				// enable/disable tap detection
				tapDetectionConfig.setYawResetEnabled(tapDetectionSettings.yawResetEnabled());
				tapDetectionConfig.setFullResetEnabled(tapDetectionSettings.fullResetEnabled());
				tapDetectionConfig
					.setMountingResetEnabled(tapDetectionSettings.mountingResetEnabled());
				tapDetectionConfig.setSetupMode(tapDetectionSettings.setupMode());

				// set number of trackers that can have high accel before taps
				// are rejected
				if (tapDetectionSettings.hasNumberTrackersOverThreshold()) {
					tapDetectionConfig
						.setNumberTrackersOverThreshold(
							tapDetectionSettings.numberTrackersOverThreshold()
						);
				}

				// set tap detection delays
				if (tapDetectionSettings.hasYawResetDelay()) {
					tapDetectionConfig.setYawResetDelay(tapDetectionSettings.yawResetDelay());
				}
				if (tapDetectionSettings.hasFullResetDelay()) {
					tapDetectionConfig.setFullResetDelay(tapDetectionSettings.fullResetDelay());
				}
				if (tapDetectionSettings.hasMountingResetDelay()) {
					tapDetectionConfig
						.setMountingResetDelay(tapDetectionSettings.mountingResetDelay());
				}

				// set the number of taps required for each action
				if (tapDetectionSettings.hasYawResetTaps()) {
					tapDetectionConfig
						.setYawResetTaps(tapDetectionSettings.yawResetTaps());
				}
				if (tapDetectionSettings.hasFullResetTaps()) {
					tapDetectionConfig
						.setFullResetTaps(tapDetectionSettings.fullResetTaps());
				}
				if (tapDetectionSettings.hasMountingResetTaps()) {
					tapDetectionConfig
						.setMountingResetTaps(tapDetectionSettings.mountingResetTaps());
				}

				this.api.server.humanPoseManager.updateTapDetectionConfig();
			}
		}

		var modelSettings = req.modelSettings();
		if (modelSettings != null) {
			var hpm = this.api.server.humanPoseManager;
			var legTweaksConfig = this.api.server.configManager.getVrConfig().getLegTweaks();
			var toggles = modelSettings.toggles();
			var ratios = modelSettings.ratios();
			var legTweaks = modelSettings.legTweaks();

			if (toggles != null) {
				// Note: toggles.has____ returns the same as toggles._____ this
				// seems like a bug
				hpm.setToggle(SkeletonConfigToggles.EXTENDED_SPINE_MODEL, toggles.extendedSpine());
				hpm
					.setToggle(
						SkeletonConfigToggles.EXTENDED_PELVIS_MODEL,
						toggles.extendedPelvis()
					);
				hpm.setToggle(SkeletonConfigToggles.EXTENDED_KNEE_MODEL, toggles.extendedKnee());
				hpm
					.setToggle(
						SkeletonConfigToggles.FORCE_ARMS_FROM_HMD,
						toggles.forceArmsFromHmd()
					);
				hpm.setToggle(SkeletonConfigToggles.EXTENDED_SPINE_MODEL, toggles.extendedSpine());
				hpm
					.setToggle(
						SkeletonConfigToggles.EXTENDED_PELVIS_MODEL,
						toggles.extendedPelvis()
					);
				hpm.setToggle(SkeletonConfigToggles.EXTENDED_KNEE_MODEL, toggles.extendedKnee());
				hpm
					.setToggle(
						SkeletonConfigToggles.FORCE_ARMS_FROM_HMD,
						toggles.forceArmsFromHmd()
					);
				hpm.setToggle(SkeletonConfigToggles.FLOOR_CLIP, toggles.floorClip());
				hpm
					.setToggle(
						SkeletonConfigToggles.SKATING_CORRECTION,
						toggles.skatingCorrection()
					);
				hpm.setToggle(SkeletonConfigToggles.VIVE_EMULATION, toggles.viveEmulation());
				hpm.setToggle(SkeletonConfigToggles.TOE_SNAP, toggles.toeSnap());
				hpm.setToggle(SkeletonConfigToggles.FOOT_PLANT, toggles.footPlant());
				hpm.setToggle(SkeletonConfigToggles.SELF_LOCALIZATION, toggles.selfLocalization());
			}

			if (ratios != null) {
				if (ratios.hasImputeWaistFromChestHip()) {
					hpm
						.setValue(
							SkeletonConfigValues.WAIST_FROM_CHEST_HIP_AVERAGING,
							Math.max(0, ratios.imputeWaistFromChestHip())
						);
				}
				if (ratios.hasImputeWaistFromChestLegs()) {
					hpm
						.setValue(
							SkeletonConfigValues.WAIST_FROM_CHEST_LEGS_AVERAGING,
							Math.max(0, ratios.imputeWaistFromChestLegs())
						);
				}
				if (ratios.hasImputeHipFromChestLegs()) {
					hpm
						.setValue(
							SkeletonConfigValues.HIP_FROM_CHEST_LEGS_AVERAGING,
							Math.max(0, ratios.imputeHipFromChestLegs())
						);
				}
				if (ratios.hasImputeHipFromWaistLegs()) {
					hpm
						.setValue(
							SkeletonConfigValues.HIP_FROM_WAIST_LEGS_AVERAGING,
							Math.max(0, ratios.imputeHipFromWaistLegs())
						);
				}
				if (ratios.hasInterpHipLegs()) {
					hpm
						.setValue(
							SkeletonConfigValues.HIP_LEGS_AVERAGING,
							Math.max(0, ratios.interpHipLegs())
						);
				}
				if (ratios.hasInterpKneeTrackerAnkle()) {
					hpm
						.setValue(
							SkeletonConfigValues.KNEE_TRACKER_ANKLE_AVERAGING,
							Math.max(0, ratios.interpKneeTrackerAnkle())
						);
				}
				if (ratios.hasInterpKneeAnkle()) {
					hpm
						.setValue(
							SkeletonConfigValues.KNEE_ANKLE_AVERAGING,
							Math.max(0, ratios.interpKneeAnkle())
						);
				}
			}

			if (legTweaks != null) {
				if (legTweaks.hasCorrectionStrength()) {
					legTweaksConfig.setCorrectionStrength(legTweaks.correctionStrength());
				}
				this.api.server.humanPoseManager.updateLegTweaksConfig();
			}

			hpm.saveConfig();

		}

		var autoBoneSettings = req.autoBoneSettings();
		if (autoBoneSettings != null) {
			AutoBoneConfig autoBoneConfig = this.api.server.configManager
				.getVrConfig()
				.getAutoBone();

			RPCSettingsBuilder.readAutoBoneSettings(autoBoneSettings, autoBoneConfig);
		}

		if (req.resetsSettings() != null) {
			ResetsConfig resetsConfig = this.api.server.configManager
				.getVrConfig()
				.getResetsConfig();
			ArmsResetModes mode = ArmsResetModes
				.fromId(Math.max(req.resetsSettings().armsMountingResetMode(), 0));
			if (mode != null) {
				resetsConfig.setMode(mode);
			}
			resetsConfig.setResetMountingFeet(req.resetsSettings().resetMountingFeet());
			resetsConfig.updateTrackersResetsSettings();
		}

		this.api.server.configManager.saveConfig();
	}

}

File: server/core/src/main/java/dev/slimevr/protocol/rpc/settings/RPCSettingsHandler.kt
package dev.slimevr.protocol.rpc.settings

import com.google.flatbuffers.FlatBufferBuilder
import dev.slimevr.bridge.ISteamVRBridge
import dev.slimevr.config.ArmsResetMode
import dev.slimevr.config.OSCConfig
import dev.slimevr.config.TapDetectionConfig
import dev.slimevr.config.VMCConfig
import dev.slimevr.config.VRCOSCConfig
import dev.slimevr.filtering.TrackerFilters
import dev.slimevr.protocol.GenericConnection
import dev.slimevr.protocol.ProtocolAPI
import dev.slimevr.protocol.rpc.RPCHandler
import dev.slimevr.tracking.processor.config.SkeletonConfigToggles
import dev.slimevr.tracking.processor.config.SkeletonConfigValues
import dev.slimevr.tracking.trackers.TrackerRole
import dev.slimevr.tracking.trackers.toKey
import solarxr_protocol.rpc.ChangeSettingsRequest
import solarxr_protocol.rpc.RpcMessage
import solarxr_protocol.rpc.RpcMessageHeader
import solarxr_protocol.rpc.SettingsResponse
import kotlin.math.*

class RPCSettingsHandler(var rpcHandler: RPCHandler, var api: ProtocolAPI) {
	init {
		rpcHandler.registerPacketListener(RpcMessage.SettingsRequest) { conn: GenericConnection, messageHeader: RpcMessageHeader? ->
			this.onSettingsRequest(
				conn,
				messageHeader,
			)
		}
		rpcHandler
			.registerPacketListener(
				RpcMessage.ChangeSettingsRequest,
			) { conn: GenericConnection?, messageHeader: RpcMessageHeader ->
				this.onChangeSettingsRequest(
					conn,
					messageHeader,
				)
			}
		rpcHandler.registerPacketListener(RpcMessage.SettingsResetRequest) { conn: GenericConnection, messageHeader: RpcMessageHeader? ->
			this.onSettingsResetRequest(conn, messageHeader)
		}
	}

	fun onSettingsRequest(conn: GenericConnection, messageHeader: RpcMessageHeader?) {
		rpcHandler.sendSettingsChangedResponse(conn, messageHeader)
	}

	fun onChangeSettingsRequest(conn: GenericConnection?, messageHeader: RpcMessageHeader) {
		val req = messageHeader
			.message(ChangeSettingsRequest()) as? ChangeSettingsRequest ?: return

		if (req.steamVrTrackers() != null) {
			val bridge = api.server
				.getVRBridge(ISteamVRBridge::class.java)

			if (bridge != null) {
				bridge.changeShareSettings(TrackerRole.WAIST, req.steamVrTrackers().waist())
				bridge.changeShareSettings(TrackerRole.CHEST, req.steamVrTrackers().chest())
				bridge.changeShareSettings(TrackerRole.LEFT_FOOT, req.steamVrTrackers().leftFoot())
				bridge.changeShareSettings(TrackerRole.RIGHT_FOOT, req.steamVrTrackers().rightFoot())
				bridge.changeShareSettings(TrackerRole.LEFT_KNEE, req.steamVrTrackers().leftKnee())
				bridge.changeShareSettings(TrackerRole.RIGHT_KNEE, req.steamVrTrackers().rightKnee())
				bridge.changeShareSettings(TrackerRole.LEFT_ELBOW, req.steamVrTrackers().leftElbow())
				bridge.changeShareSettings(TrackerRole.RIGHT_ELBOW, req.steamVrTrackers().rightElbow())
				bridge.changeShareSettings(TrackerRole.LEFT_HAND, req.steamVrTrackers().leftHand())
				bridge.changeShareSettings(TrackerRole.RIGHT_HAND, req.steamVrTrackers().rightHand())
				bridge.setAutomaticSharedTrackers(req.steamVrTrackers().automaticTrackerToggle())
			}
		}

		if (req.filtering() != null) {
			val type = TrackerFilters.fromId(req.filtering().type())
			if (type != null) {
				api.server.configManager.settings.update {
					it.copy(
						filters = it.filters.copy(
							type = type.configKey,
							amount = req.filtering().amount(),
						),
					)
				}
				filtersConfig.updateTrackersFilters() // FIXME: move this outside of the config file
			}
		}

		if (req.oscRouter() != null) {
			val oscRouter = api.server.oSCRouter
			val osc = req.oscRouter().oscSettings()
			api.server.configManager.settings.update {
				it.copy(
					oscRouter = if (osc != null) {
						OSCConfig(
							enabled = osc.enabled(),
							portIn = osc.portIn(),
							portOut = osc.portOut(),
							address = osc.address(),
						)
					} else {
						it.oscRouter
					},
				)
			}
			oscRouter.refreshSettings(true)
		}

		if (req.vrcOsc() != null) {
			val osc = req.vrcOsc().oscSettings()
			val trackers = req.vrcOsc().trackers()
			api.server.configManager.settings.update {
				it.copy(
					vrcOSC = if (osc != null) {
						VRCOSCConfig(
							enabled = osc.enabled(),
							portIn = osc.portIn(),
							portOut = osc.portOut(),
							address = osc.address(),
							oscqueryEnabled = req.vrcOsc().oscqueryEnabled(),
							trackers = if (trackers != null) {
								mutableMapOf(
									TrackerRole.HEAD.toKey to trackers.head(),
									TrackerRole.CHEST.toKey to trackers.chest(),
									TrackerRole.WAIST.toKey to trackers.waist(),
									TrackerRole.LEFT_KNEE.toKey to trackers.knees(),
									TrackerRole.RIGHT_KNEE.toKey to trackers.knees(),
									TrackerRole.LEFT_FOOT.toKey to trackers.feet(),
									TrackerRole.RIGHT_FOOT.toKey to trackers.feet(),
									TrackerRole.LEFT_ELBOW.toKey to trackers.elbows(),
									TrackerRole.RIGHT_ELBOW.toKey to trackers.elbows(),
									TrackerRole.LEFT_HAND.toKey to trackers.hands(),
									TrackerRole.RIGHT_HAND.toKey to trackers.hands(),
								)
							} else {
								it.vrcOSC.trackers
							},
						)
					} else {
						it.vrcOSC
					},
				)
			}
			val vrcOscHandler = api.server.vrcOSCHandler
			vrcOscHandler.refreshSettings(true)
		}

		if (req.vmcOsc() != null) {
			val vmcHandler = api.server.vMCHandler
			val osc = req.vmcOsc().oscSettings()
			api.server.configManager.settings.update {
				it.copy(
					vmc = if (osc != null) {
						VMCConfig(
							enabled = osc.enabled(),
							portIn = osc.portIn(),
							portOut = osc.portOut(),
							address = osc.address(),
							vrmJson = if (req.vmcOsc().vrmJson() != null) req.vmcOsc().vrmJson().ifEmpty { null } else it.vmc.vrmJson,
							anchorHip = req.vmcOsc().anchorHip(),
							mirrorTracking = req.vmcOsc().mirrorTracking(),
						)
					} else {
						it.vmc
					},
				)
			}
			vmcHandler.refreshSettings(true)
		}

		if (req.tapDetectionSettings() != null) {
			val tapDetectionSettings = req.tapDetectionSettings()
			if (tapDetectionSettings != null) {
				api.server.configManager.settings.update {
					it.copy(
						tapDetection = TapDetectionConfig(
							yawResetEnabled = tapDetectionSettings.yawResetEnabled(),
							fullResetEnabled = tapDetectionSettings.fullResetEnabled(),
							mountingResetEnabled = tapDetectionSettings.mountingResetEnabled(),
							setupMode = tapDetectionSettings.setupMode(),
							numberTrackersOverThreshold = tapDetectionSettings.numberTrackersOverThreshold(),
							yawResetDelay = tapDetectionSettings.yawResetDelay(),
							fullResetDelay = tapDetectionSettings.fullResetDelay(),
							mountingResetDelay = tapDetectionSettings.mountingResetDelay(),
							yawResetTaps = tapDetectionSettings.yawResetTaps(),
							fullResetTaps = tapDetectionSettings.fullResetTaps(),
							mountingResetTaps = tapDetectionSettings.mountingResetTaps(),
						)
					)
				}
				api.server.humanPoseManager.updateTapDetectionConfig()
			}
		}

		val modelSettings = req.modelSettings()
		if (modelSettings != null) {
			val hpm = api.server.humanPoseManager
			val toggles = modelSettings.toggles()
			val ratios = modelSettings.ratios()
			val legTweaks = modelSettings.legTweaks()

			if (toggles != null) {
				// Note: toggles.has____ returns the same as toggles._____ this
				// seems like a bug
				hpm.setToggle(SkeletonConfigToggles.EXTENDED_SPINE_MODEL, toggles.extendedSpine())
				hpm
					.setToggle(
						SkeletonConfigToggles.EXTENDED_PELVIS_MODEL,
						toggles.extendedPelvis(),
					)
				hpm.setToggle(SkeletonConfigToggles.EXTENDED_KNEE_MODEL, toggles.extendedKnee())
				hpm
					.setToggle(
						SkeletonConfigToggles.FORCE_ARMS_FROM_HMD,
						toggles.forceArmsFromHmd(),
					)
				hpm.setToggle(SkeletonConfigToggles.EXTENDED_SPINE_MODEL, toggles.extendedSpine())
				hpm
					.setToggle(
						SkeletonConfigToggles.EXTENDED_PELVIS_MODEL,
						toggles.extendedPelvis(),
					)
				hpm.setToggle(SkeletonConfigToggles.EXTENDED_KNEE_MODEL, toggles.extendedKnee())
				hpm
					.setToggle(
						SkeletonConfigToggles.FORCE_ARMS_FROM_HMD,
						toggles.forceArmsFromHmd(),
					)
				hpm.setToggle(SkeletonConfigToggles.FLOOR_CLIP, toggles.floorClip())
				hpm
					.setToggle(
						SkeletonConfigToggles.SKATING_CORRECTION,
						toggles.skatingCorrection(),
					)
				hpm.setToggle(SkeletonConfigToggles.TOE_SNAP, toggles.toeSnap())
				hpm.setToggle(SkeletonConfigToggles.FOOT_PLANT, toggles.footPlant())
				hpm.setToggle(SkeletonConfigToggles.SELF_LOCALIZATION, toggles.selfLocalization())
				hpm.setToggle(SkeletonConfigToggles.USE_POSITION, toggles.usePosition())
				hpm.setToggle(SkeletonConfigToggles.ENFORCE_CONSTRAINTS, toggles.enforceConstraints())
				hpm.setToggle(SkeletonConfigToggles.CORRECT_CONSTRAINTS, toggles.correctConstraints())
			}

			if (ratios != null) {
				if (ratios.hasImputeWaistFromChestHip()) {
					hpm
						.setValue(
							SkeletonConfigValues.WAIST_FROM_CHEST_HIP_AVERAGING,
							max(0f, ratios.imputeWaistFromChestHip()),
						)
				}
				if (ratios.hasImputeWaistFromChestLegs()) {
					hpm
						.setValue(
							SkeletonConfigValues.WAIST_FROM_CHEST_LEGS_AVERAGING,
							max(0f, ratios.imputeWaistFromChestLegs()),
						)
				}
				if (ratios.hasImputeHipFromChestLegs()) {
					hpm
						.setValue(
							SkeletonConfigValues.HIP_FROM_CHEST_LEGS_AVERAGING,
							max(0f, ratios.imputeHipFromChestLegs()),
						)
				}
				if (ratios.hasImputeHipFromWaistLegs()) {
					hpm
						.setValue(
							SkeletonConfigValues.HIP_FROM_WAIST_LEGS_AVERAGING,
							max(0f, ratios.imputeHipFromWaistLegs()),
						)
				}
				if (ratios.hasInterpHipLegs()) {
					hpm
						.setValue(
							SkeletonConfigValues.HIP_LEGS_AVERAGING,
							max(0f, ratios.interpHipLegs()),
						)
				}
				if (ratios.hasInterpKneeTrackerAnkle()) {
					hpm
						.setValue(
							SkeletonConfigValues.KNEE_TRACKER_ANKLE_AVERAGING,
							max(0f, ratios.interpKneeTrackerAnkle()),
						)
				}
				if (ratios.hasInterpKneeAnkle()) {
					hpm
						.setValue(
							SkeletonConfigValues.KNEE_ANKLE_AVERAGING,
							max(0f, ratios.interpKneeAnkle()),
						)
				}
			}


			if (legTweaks != null) {
				if (legTweaks.hasCorrectionStrength()) {
					api.server.configManager.settings.update {
						it.copy(legTweaks = it.legTweaks.copy(
							correctionStrength = legTweaks.correctionStrength()
						))
					}
					api.server.humanPoseManager.updateLegTweaksConfig()
				}
			}

			modelSettings.skeletonHeight()?.let { skeletonHeight ->
				api.server.configManager.user.update {
					it.copy(
						skeleton = it.skeleton.copy(
							hmdHeight = skeletonHeight.hmdHeight(),
							floorHeight = skeletonHeight.floorHeight()
						)
					)
				}
			}

			hpm.saveConfig()
		}

		val autoBoneSettings = req.autoBoneSettings()
		if (autoBoneSettings != null) {
			val autoBoneConfig = api.server.configManager
				.vrConfig
				.autoBone

			RPCSettingsBuilder.readAutoBoneSettings(autoBoneSettings, autoBoneConfig)
		}

		if (req.resetsSettings() != null) {
			val resetsConfig = api.server.configManager
				.vrConfig
				.resetsConfig
			val mode = ArmsResetMode
				.fromId(max(req.resetsSettings().armsMountingResetMode(), 0))
			if (mode != null) {
				resetsConfig.mode = mode
			}
			resetsConfig.resetMountingFeet = req.resetsSettings().resetMountingFeet()
			resetsConfig.saveMountingReset = req.resetsSettings().saveMountingReset()
			resetsConfig.yawResetSmoothTime = req.resetsSettings().yawResetSmoothTime()
			resetsConfig.resetHmdPitch = req.resetsSettings().resetHmdPitch()
			resetsConfig.updateTrackersResetsSettings()
		}

		if (req.stayAligned() != null) {
			val config = api.server.configManager.vrConfig.stayAlignedConfig
			val requestConfig = req.stayAligned()
			config.enabled = requestConfig.enabled()
			config.hideYawCorrection = requestConfig.hideYawCorrection()
			config.standingRelaxedPose.enabled = requestConfig.standingEnabled()
			config.standingRelaxedPose.upperLegAngleInDeg = requestConfig.standingUpperLegAngle()
			config.standingRelaxedPose.lowerLegAngleInDeg = requestConfig.standingLowerLegAngle()
			config.standingRelaxedPose.footAngleInDeg = requestConfig.standingFootAngle()
			config.sittingRelaxedPose.enabled = requestConfig.sittingEnabled()
			config.sittingRelaxedPose.upperLegAngleInDeg = requestConfig.sittingUpperLegAngle()
			config.sittingRelaxedPose.lowerLegAngleInDeg = requestConfig.sittingLowerLegAngle()
			config.sittingRelaxedPose.footAngleInDeg = requestConfig.sittingFootAngle()
			config.flatRelaxedPose.enabled = requestConfig.flatEnabled()
			config.flatRelaxedPose.upperLegAngleInDeg = requestConfig.flatUpperLegAngle()
			config.flatRelaxedPose.lowerLegAngleInDeg = requestConfig.flatLowerLegAngle()
			config.flatRelaxedPose.footAngleInDeg = requestConfig.flatFootAngle()
		}

		if (req.hidSettings() != null) {
			val config = api.server.configManager.vrConfig.hidConfig
			val requestConfig = req.hidSettings()
			config.trackersOverHID = requestConfig.trackersOverHid()
		}

		api.server.configManager.settings.save()
		api.server.configManager.user.save()
	}

	fun onSettingsResetRequest(conn: GenericConnection, messageHeader: RpcMessageHeader?) {
		api.server.configManager.resetConfig()
	}

	companion object {
		fun sendSteamVRUpdatedSettings(api: ProtocolAPI, rpcHandler: RPCHandler) {
			val fbb = FlatBufferBuilder(32)
			val bridge: ISteamVRBridge =
				api.server.getVRBridge(ISteamVRBridge::class.java) ?: return

			val settings = SettingsResponse
				.createSettingsResponse(
					fbb,
					RPCSettingsBuilder.createSteamVRSettings(fbb, bridge), 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
				)
			val outbound =
				rpcHandler.createRPCMessage(fbb, RpcMessage.SettingsResponse, settings)
			fbb.finish(outbound)
			api.apiServers.forEach { apiServer ->
				apiServer.apiConnections.forEach { it.send(fbb.dataBuffer()) }
			}
		}
	}
}
